### PR TITLE
Simulation Stack - January Release

### PR DESCRIPTION
**Pull request by iomaganaris**
**New package versions:**
- Neuron 7.8.0b: Fix the load balance issue
- Synapsetool 0.5.8: consolidated indexing, prevent truncating, improved clang compat
- neurodamus-core 2.9.1: Delete corenrn data, fixed MPI guards, improved clang compat
- neurodamus-py 0.8.0: new CLI options: keep corenrn data (default delete), build model only...
- reportinglib 2.5.3: Improved MPI 2+ compat (drop deprecated calls)

**Other recipe changes:**
- Deployment of neurodamus-core as a standalone compiled module.
 Initial support for any neurodmus-core + model-xxx for neurodamus-py 
- neurodamus-core, neurodamus-model-xxx, sim-model-xxx all to share the same code to build mod libraries with support for core-neuron and auto detection of dependencies.
- Pin synapsetool for older versions of parquet-converters

**Extra**
- deploy spatial-index v0.2.1

### CODE REVIEW

**Comment by matz-e** [comment]
Please include #650, too?

**Comment by ferdonline** [comment]
rebased and included my commits

File: var/spack/repos/builtin/packages/model-neocortex/package.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack import *
+from spack.pkg.builtin.sim_model import SimModel
+
+
+class ModelNeocortex(SimModel):
+    """
+    The Norcortex neuron mechanisms for (core)Neuron
```
Comment (iomaganaris) [inline var/spack/repos/builtin/packages/model-neocortex/package.py]: Nitpick.. Spelling error in `Nocortex`

**Comment by ferdonline** [comment]
please retest

**Review by ferdonline — COMMENTED** [review]
I am done with changes. Also this setup is tested in Simulation Stack CI (which we shall not forget anymore to avoid problems like last time with Neuron 😞) https://bbpcode.epfl.ch/ci/blue/organizations/jenkins/hpc.SimulationStack/detail/hpc.SimulationStack/1041/pipeline
Please review

File: deploy/packages/bbp-packages.yaml [hunk 2]
```diff
@@ -30,7 +30,7 @@ packages:
       - brainbuilder@0.11.7
       - brion@3.0.0 +python ^boost@1.65.0
       - connectome-tools@0.3.2
-      - spatial-index@0.1.0
+      - spatial-index@0.2.1
```
Comment (matz-e) [inline deploy/packages/bbp-packages.yaml:33]: Is this part of the simulation stack? Wasn't aware of that before…

File: deploy/configs/applications/modules.yaml [hunk 3]
```diff
@@ -39,8 +39,7 @@ modules:
       - functionalizer
       - meshball
       - nest
-      - 'neurodamus%intel'
-      - 'neurodamus-core+common'
+      - neurodamus-core
```
Comment (pramodk) [inline deploy/configs/applications/modules.yaml:43]: to clarify: Are we deploying +common? If so, Isn't `+common` better to be sure that neurodamus-core with +common is deployed?

File: deploy/packages/bbp-packages.yaml [hunk 4]
```diff
@@ -173,6 +173,7 @@ packages:
       - mpi
       - python
     specs:
+      - neurodamus-core
```
Comment (pramodk) [inline deploy/packages/bbp-packages.yaml]: to clarify : Is this +common or ~common?  (on line 154 we are specifying variant)

File: var/spack/repos/builtin/packages/highfive/package.py [hunk 5]
```diff
@@ -42,12 +42,13 @@ class Highfive(CMakePackage):
...
     def cmake_args(self):
         return [
-            '-DUSE_EIGEN:Bool=' + ('TRUE' if '+eigen' in self.spec else 'FALSE'),
-            '-DUSE_XTENSOR:Bool=' + ('TRUE' if '+xtensor' in self.spec else 'FALSE'),
-            '-DUSE_BOOST:Bool={0}'.format('+boost' in self.spec),
-            '-DHIGHFIVE_PARALLEL_HDF5:Bool={0}'.format('+mpi' in self.spec),
-            '-DHIGHFIVE_EXAMPLES:Bool={0}'.format(self.spec.satisfies('@develop')),
-            '-DHIGHFIVE_UNIT_TESTS:Bool={0}'.format(self.spec.satisfies('@develop')),
-            '-DHIGHFIVE_TEST_SINGLE_INCLUDES:Bool={0}'.format(self.spec.satisfies('@develop')),
+            '-DUSE_BOOST:Bool=' + str(self.spec.satisfies('+boost')),
```
Comment (pramodk) [inline var/spack/repos/builtin/packages/highfive/package.py:45]: just nitpick: its true that `self.spec.satisfies` return true/false but older syntax ('TRUE' if '+eigen' in self.spec else 'FALSE') is common in spack packages.

File: var/spack/repos/builtin/packages/neuron/package.py [hunk 6]
```diff
@@ -25,9 +25,14 @@ class Neuron(Package):
 
     # Patch which reverts 81a7a39 for numerical compat
     patch('revert_Import3d_numerical_format.patch', when='@7.8.0:')
+    # Patch which reverts d9605cb for not hanging on ExperimentalMechComplex
+    patch('revert_d9605cb.patch', when='@7.8.0a')
+
+    # Patch which applies load balancing fix
+    patch('apply_79a4d2af_load_balance_fix.patch', when='@7.8.0:')
```
Comment (matz-e) [inline var/spack/repos/builtin/packages/neuron/package.py:32]: The patches look nearly identical to me. Any chance we could re-use the same patch? Or at least name them similarly, so that it is clear that they are for different versions?

File: var/spack/repos/builtin/packages/synapsetool/package.py [hunk 7]
```diff
@@ -34,6 +34,8 @@ class Synapsetool(CMakePackage):
     git      = "ssh://bbpcode.epfl.ch/hpc/synapse-tool"
 
     version('develop', submodules=True)
+    version('0.5.8', tag='v0.5.8', submodules=True)
+    version('0.5.7', tag='v0.5.7', submodules=True)
```
Comment (pramodk) [inline var/spack/repos/builtin/packages/synapsetool/package.py]: If `0.5.7` is not published/released, can it be dropped?

File: deploy/configs/applications/modules.yaml (hunk 3 above)
Comment (ferdonline) [reply deploy/configs/applications/modules.yaml:43]: We allow both.
`+common` is deployed just because of bglibpy and Werner case. In the generic case people will be loading a pure core (~common) and the cell mechanisms they want, including their own common. That 's the case of the pre-build model-neocortex, and others to come.

File: var/spack/repos/builtin/packages/py-neurodamus/package.py [hunk 8]
```diff
@@ -23,3 +24,13 @@ class PyNeurodamus(PythonPackage):
...
     depends_on('py-six',           type='run')
+
+    @run_after('install')
+    def install_scripts(self):
+        mkdirp(self.prefix.share)
+        for script in ('init.py', '_debug.py'):
+            copy(script, self.prefix.share)
+
+    def setup_environment(self, spack_env, run_env):
+        PythonPackage.setup_environment(self, spack_env, run_env)
```
Comment (matz-e) [inline var/spack/repos/builtin/packages/py-neurodamus/package.py:35]: Good call, but a heads up: with the new deployment, `setup_environment` is deprecated. Packages upstream may migrate to `setup_{run,build}_environment`, so the parent method is not guaranteed to exist.

Nothing urgent here, just a note to myself and FYI.

File: deploy/configs/applications/modules.yaml (hunk 3 above)
Comment (ferdonline) [reply deploy/configs/applications/modules.yaml:43]: Importantly, `core+common` cannot be used together with `model-neocortex` as both would bring the common (AMPANMDA, gap, etc)

File: var/spack/repos/builtin/packages/model-neocortex/package.py [hunk 9]
```diff
@@ -0,0 +1,35 @@
+# Copyright 2013-2018 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack import *
+from spack.pkg.builtin.sim_model import SimModel
+
+
+class ModelNeocortex(SimModel):
```
Comment (pramodk) [inline var/spack/repos/builtin/packages/model-neocortex/package.py:10]: where is this being used? at least I don't find it's usage in this PR (?)

File: var/spack/repos/builtin/packages/sim-model/package.py [hunk 10]
```diff
@@ -37,61 +42,67 @@ class ModelHippocampus(SimModel):
...
-            mechlib = self.__build_mods_coreneuron(
-                corenrn_mods or mods_location, link_flag, include_flag)
-            # Link neuron special with this mechs lib
-            link_flag += ' %s %s' % (mechlib.ld_flags, self._get_link_flags('coreneuron'))
+            libnrncoremech = self.__build_mods_coreneuron(
+                corenrn_mods or mods_location, link_flag, include_flag
+            )
+            # Relevant flags to build neuron's nrnmech lib
             include_flag += ' -DENABLE_CORENEURON'  # only now, otherwise mods assume neuron
+            include_flag += ' -I%s' % self.spec['coreneuron'].prefix.include
```
Comment (matz-e) [inline var/spack/repos/builtin/packages/sim-model/package.py:75]: Could this not just be string concatenation?

File: var/spack/repos/builtin/packages/sim-model/package.py [hunk 11]
```diff
@@ -100,34 +111,40 @@ def install(self, spec, prefix, install_src=True):
...
+                elif self.spec.satisfies('^coreneuron@0.17:'):
+                    which('nrnivmodl-core')("-d", prefix)  # Set dest to install
+
+        # Install special
         shutil.copy(join_path(arch, 'special'), prefix.bin)
 
         if self.spec.satisfies('^neuron~binary'):
             # Install libnrnmech - might have several links.
-            for f in find(arch + "/.libs", 'libnrnmech*.so*', recursive=False):
+            for f in find(arch + '/.libs', 'libnrnmech*.so*', recursive=False):
```
Comment (matz-e) [inline var/spack/repos/builtin/packages/sim-model/package.py:137]: ' -> " makes the diff harder to read…

File: var/spack/repos/builtin/packages/sim-model/package.py [hunk 12]
```diff
@@ -139,23 +156,34 @@ def _install_src(self, prefix):
...
-        for cmod in find(arch, "*.c", recursive=False):
+        for cmod in find(arch, '*.c', recursive=False):
             shutil.move(cmod, prefix.share.modc)
 
-        if self.spec.satisfies('+coreneuron'):
-            shutil.move(join_path(arch, 'core/mod2c'), prefix.share.modc_core)
+    def _setup_environment_common(self, spack_env, run_env):
+        spack_env.unset('LC_ALL')
+        # Remove LD_LIB_PATHs
+        to_rem = ('LD_LIBRARY_PATH', 'DYLD_LIBRARY_PATH', 'DYLD_FALLBACK_LIBRARY_PATH')
```
Comment (matz-e) [inline var/spack/repos/builtin/packages/sim-model/package.py]: `to_rem` is ambiguous. I read it as `to_remember`. Please change to `to_rm` or other.

File: var/spack/repos/builtin/packages/neurodamus-core/package.py [hunk 13]
```diff
@@ -1,103 +1,112 @@
...
+    variant('synapsetool',  default=True, description="Enable SynapseTool reader (for edges)")
+    variant('mvdtool',      default=True, description="Enable MVDTool reader (for nodes)")
+
+    # NOTE: Several variants / dependencies come from SimModel
+    depends_on("mpi",  when='+mpi', type=('build', 'run'))  # dont link
+    depends_on("hdf5+mpi", when='+hdf5+mpi')
+    depends_on("hdf5~mpi", when='+hdf5~mpi')
+    depends_on('reportinglib',         when='+reportinglib')
+    depends_on('reportinglib+profile', when='+reportinglib+profile')
+    depends_on('synapsetool',          when='+synapsetool')
```
Comment (pramodk) [inline var/spack/repos/builtin/packages/neurodamus-core/package.py:49]: why reportinglib is dependency with neurodamus-core? IIRC, neurodamus-core was not building auxiliary mod files (report, hdf5 etc related).

similar question for synapsetool, mvdtool etc.

**Review by matz-e — APPROVED** [review]
Some comments, but nothing that should stop this from going in.

File: var/spack/repos/builtin/packages/neurodamus-core/package.py [hunk 14]
```diff
@@ -1,103 +1,112 @@
...
+    depends_on('py-mvdtool',           when='+mvdtool', type='run')
+    # If external & static, we must bring their dependencies.
+    depends_on('zlib')  # for hdf5
```
Comment (pramodk) [inline var/spack/repos/builtin/packages/neurodamus-core/package.py:52]: nitpick : I see that zlib is dependency of hdf5 but wondering in what context you saw link error.

File: var/spack/repos/builtin/packages/neurodamus-core/package.py [hunk 15]
```diff
@@ -1,103 +1,112 @@
...
-            if os.path.isfile(bindir_info):
-                bindir = open(bindir_info, 'r').readline()
-                mechlib = find_libraries('libnrnmech', join_path(self.prefix, bindir, '.libs'))[0]
-                run_env.set('NRNMECH_LIB_PATH', mechlib)
-                run_env.set('BGLIBPY_MOD_LIBRARY_PATH', mechlib)
-            else:
-                tty.warn("No .bindir info file found. NRNMECH_LIB_PATH env var wont be set")
+        self._setup_environment_common(spack_env, run_env)
+        for lib in find(self.prefix.lib, 'libnrnmech*'):
+            run_env.set('NRNMECH_LIB_PATH', lib)
```
Comment (pramodk) [inline var/spack/repos/builtin/packages/neurodamus-core/package.py:114]: if we are `setting single` env var, how this works in loop?

File: var/spack/repos/builtin/packages/neurodamus-model/package.py [hunk 16]
```diff
@@ -19,15 +19,13 @@ class NeurodamusModel(SimModel):
...
                                                       " other string: alternate path")
 
-    depends_on('neurodamus-core')
-    depends_on('neurodamus-core@develop', when='@develop')
-    depends_on('neurodamus-core+python', when='+python')
+    depends_on('neurodamus-core', type='build')
+    depends_on('neurodamus-core@develop', type='build', when='@develop')
 
-    depends_on("mpi")
+    depends_on("mpi", type=('build', 'run'))
```
Comment (pramodk) [inline var/spack/repos/builtin/packages/neurodamus-model/package.py]: mpi is typically link as well.
```suggestion
    depends_on("mpi")
```

MPI is also typically link dependency.

File: var/spack/repos/builtin/packages/neurodamus-model/package.py [hunk 17]
```diff
@@ -112,50 +101,39 @@ def install(self, spec, prefix):
...
-            run_env.set('NEURODAMUS_PYTHON', pylib)
+        self._setup_environment_common(spack_env, run_env)
+        for libnrnmech_name in find(self.prefix.lib, 'libnrnmech*.so', recursive=False):
+            # We have the two libs and must export them in different vars
+            #  - NRNMECH_LIB_PATH the combined lib (used by neurodamus-py)
+            #  - BGLIBPY_MOD_LIBRARY_PATH is the pure mechanism (used by bglib-py)
+            if '_nd.' in libnrnmech_name:
+                run_env.set('NRNMECH_LIB_PATH', libnrnmech_name)
+            else:
+                run_env.set('BGLIBPY_MOD_LIBRARY_PATH', libnrnmech_name)
```
Comment (pramodk) [inline var/spack/repos/builtin/packages/neurodamus-model/package.py:136]: can we avoid use of loop here?

File: var/spack/repos/builtin/packages/sim-model/package.py [hunk 18]
```diff
@@ -27,8 +31,9 @@ class ModelHippocampus(SimModel):
     variant('coreneuron',  default=False, description="Enable CoreNEURON Support")
     variant('profile',     default=False, description="Enable profiling using Tau")
 
-    depends_on('neuron+mpi', type=('build', 'link', 'run'))
-    depends_on('coreneuron', when='+coreneuron', type=('build', 'link', 'run'))
+    # We dont link automatically to neuron/corenrn, nrnivmodl does it for us (=> no 'link' mode)
+    depends_on('neuron+mpi', type=('build', 'run'))
```
Comment (pramodk) [inline var/spack/repos/builtin/packages/sim-model/package.py:36]:
```suggestion
    depends_on('neuron+mpi')
```
even we dont link explicitly, they are just hidden inside script...so more clear to keep that?

File: var/spack/repos/builtin/packages/sim-model/package.py [hunk 19]
```diff
@@ -27,8 +31,9 @@ class ModelHippocampus(SimModel):
...
+    depends_on('coreneuron', when='+coreneuron', type=('build', 'run'))
```
Comment (pramodk) [inline var/spack/repos/builtin/packages/sim-model/package.py:37]: same

File: var/spack/repos/builtin/packages/sim-model/package.py [hunk 20]
```diff
@@ -37,61 +42,67 @@ class ModelHippocampus(SimModel):
...
+        return ('_' + self.mech_name) if self.mech_name else ''
 
     def _build_mods(self, mods_location, link_flag='', include_flag='', corenrn_mods=None):
         """Build shared lib & special from mods in a given path
         """
-        profile_flag = ' -DENABLE_TAU_PROFILER' if '+profile' in self.spec else ''
-        include_flag += profile_flag
+        # pass include and link flags for all dependency libraries
+        # Compiler wrappers are not used to have a more reproducible building
+        dep_names = set(self.spec.dependencies_dict('link').keys())
```
Comment (pramodk) [inline var/spack/repos/builtin/packages/sim-model/package.py]: are we diving to deep here? I dont see any package using `dependencies_dict('link')`.

File: var/spack/repos/builtin/packages/sim-model/package.py [hunk 21]
```diff
@@ -37,61 +42,67 @@ class ModelHippocampus(SimModel):
...
+        for dep in dep_names:
+            if self.spec[dep].prefix in ('/usr', '/usr/local'):
+                raise Exception('Dependency lib "' + dep + '" coming from system prefix')
```
Comment (pramodk) [inline var/spack/repos/builtin/packages/sim-model/package.py]: why exception? we often use dependencies like hdf5 from /usr.

File: var/spack/repos/builtin/packages/sim-model/package.py [hunk 22]
```diff
@@ -37,61 +42,67 @@ class ModelHippocampus(SimModel):
...
+        link_flag += ' ' + ' '.join(self.spec[dep].libs.ld_flags for dep in dep_names)
+        link_flag += ' ' + ' '.join(self.spec[dep].libs.rpath_flags for dep in dep_names)
```
Comment (pramodk) [inline var/spack/repos/builtin/packages/sim-model/package.py]: Is this not indented correctly? (should be inside loop?)

File: var/spack/repos/builtin/packages/model-neocortex/package.py (hunk 9 above)
Comment (iomaganaris) [inline var/spack/repos/builtin/packages/model-neocortex/package.py:10]: Shouldn't there be a similar recipe for hIppocampus, mousify and thalamus models?

File: var/spack/repos/builtin/packages/sim-model/package.py [hunk 23]
```diff
@@ -139,23 +156,34 @@ def _install_src(self, prefix):
...
             shutil.move(cmod, prefix.share.modc)
 
-        if self.spec.satisfies('+coreneuron'):
-            shutil.move(join_path(arch, 'core/mod2c'), prefix.share.modc_core)
+    def _setup_environment_common(self, spack_env, run_env):
+        spack_env.unset('LC_ALL')
+        # Remove LD_LIB_PATHs
+        to_rem = ('LD_LIBRARY_PATH', 'DYLD_LIBRARY_PATH', 'DYLD_FALLBACK_LIBRARY_PATH')
+        run_env.env_modifications = [envmod for envmod in run_env.env_modifications
+                                     if envmod.name not in to_rem]
```
Comment (pramodk) [inline var/spack/repos/builtin/packages/sim-model/package.py]: Why is this required?

File: var/spack/repos/builtin/packages/sim-model/package.py [hunk 24]
```diff
@@ -139,23 +156,34 @@ def _install_src(self, prefix):
...
+        if os.path.isdir(self.prefix.lib.python):
+            run_env.prepend_path('PYTHONPATH', self.prefix.lib.python)
 
     def setup_environment(self, spack_env, run_env):
-        spack_env.unset('LC_ALL')
-        run_env.set('NRNMECH_LIB_PATH', self.spec.prefix.lib.join('libnrnmech.so'))
-        run_env.set('BGLIBPY_MOD_LIBRARY_PATH', self.spec.prefix.lib.join('libnrnmech.so'))
+        self._setup_environment_common(spack_env, run_env)
+        for libnrnmech_name in find(self.prefix.lib, 'libnrnmech*.so', recursive=False):
+            run_env.prepend_path('BGLIBPY_MOD_LIBRARY_PATH', libnrnmech_name)
```
Comment (pramodk) [inline var/spack/repos/builtin/packages/sim-model/package.py:180]: Can BGLIBPY_MOD_LIBRARY_PATH have multiple paths?

File: deploy/packages/bbp-packages.yaml (hunk 2 above)
Comment (ferdonline) [reply deploy/packages/bbp-packages.yaml:33]: No, just to avoid opening a separate PR

File: var/spack/repos/builtin/packages/neuron/package.py (hunk 6 above)
Comment (ferdonline) [reply var/spack/repos/builtin/packages/neuron/package.py:32]: Patches are automatically generated, and serve different purposes. And Im too afraid of doing any manual change or marge

File: var/spack/repos/builtin/packages/sim-model/package.py (hunk 10 above)
Comment (ferdonline) [reply var/spack/repos/builtin/packages/sim-model/package.py:75]: I'm taking line length seriously now :) and splitting concatenation in some lines is ugly... it's always ugly anyway, so hope you dont mind this choice

File: var/spack/repos/builtin/packages/sim-model/package.py (hunk 11 above)
Comment (ferdonline) [reply var/spack/repos/builtin/packages/sim-model/package.py:137]: attempt at standardizing... but work is still not complete, so sorry it came mixed here, took the chance since touched a lot this file anyway.

File: var/spack/repos/builtin/packages/sim-model/package.py (hunk 12 above)
Comment (ferdonline) [reply var/spack/repos/builtin/packages/sim-model/package.py]: dont like my var names? 😄  (even though to_rm sounds better, code is not too complex to understand, theres even a comment :) )

File: deploy/packages/bbp-packages.yaml (hunk 4 above)
Comment (ferdonline) [reply deploy/packages/bbp-packages.yaml]: `~common`

File: var/spack/repos/builtin/packages/synapsetool/package.py (hunk 7 above)
Comment (ferdonline) [reply var/spack/repos/builtin/packages/synapsetool/package.py]: I believe so...

**Review by iomaganaris — COMMENTED** [review]
LGTM and works fine when I tested it (with the help of @ferdonline ). I just want to point out that `model-neocortex` + `neurodamus-core+coreneuron` is not working.

File: var/spack/repos/builtin/packages/model-neocortex/package.py (hunk 9 above)
Comment (ferdonline) [reply var/spack/repos/builtin/packages/model-neocortex/package.py:10]: It's a module to be avail to the user

File: var/spack/repos/builtin/packages/neurodamus-core/package.py (hunk 13 above)
Comment (ferdonline) [reply var/spack/repos/builtin/packages/neurodamus-core/package.py:49]: humm... I guess I failed to explain the idea. neurodamus-core will have the mods compiled, since we can load two libraries and have the core independent from the model. We hopefully in the future can drop the mixed-together `neurodamus-model`

File: var/spack/repos/builtin/packages/sim-model/package.py (hunk 20 above)
Comment (ferdonline) [reply var/spack/repos/builtin/packages/sim-model/package.py]: I crawled to source and didn't see a cleaner way of doing it. Most packages use the wrappers, we have a pretty special case IHO

File: var/spack/repos/builtin/packages/sim-model/package.py (hunk 21 above)
Comment (ferdonline) [reply var/spack/repos/builtin/packages/sim-model/package.py]: I can remove this, but I think it's pretty unusual for our central neurodamus deployments to use /usr libs, and it already caught me undesirable situations

Comment (iomaganaris) [reply var/spack/repos/builtin/packages/sim-model/package.py]: I guess that it is an issue when installed in our local systems. For example I had this issue with `zlib` which was pointing to the `/usr` folder and was also set as `Buildable: False` in my `.spack/packages.yaml`, so I had to change this as well. I don't know if other users have the `packages.yaml` file and this exception can create problems for them

File: var/spack/repos/builtin/packages/highfive/package.py (hunk 5 above)
Comment (ferdonline) [reply var/spack/repos/builtin/packages/highfive/package.py:45]: I can make a revision tonight to address all the comments since there are several small things

File: var/spack/repos/builtin/packages/neurodamus-core/package.py (hunk 14 above)
Comment (ferdonline) [reply var/spack/repos/builtin/packages/neurodamus-core/package.py:52]: When hdf5 is external and static. Hence the comment

File: var/spack/repos/builtin/packages/neurodamus-core/package.py (hunk 15 above)
Comment (ferdonline) [reply var/spack/repos/builtin/packages/neurodamus-core/package.py:114]: We will find 0 (build time) or 1 library (module creation time) so the for is a quick IF with var storage (before python 3.8 :)

File: var/spack/repos/builtin/packages/neurodamus-model/package.py (hunk 16 above)
Comment (ferdonline) [reply var/spack/repos/builtin/packages/neurodamus-model/package.py]: In our setup it was problematic, we cant do `spec['mpi'].libs (maybe something to be fixed)
But since we now use neuron with Dynamic MPI, isn’t this the desired way?

File: var/spack/repos/builtin/packages/sim-model/package.py (hunk 21 above)
Comment (ferdonline) [reply var/spack/repos/builtin/packages/sim-model/package.py]: Fair enough. This was mostly to have the dependencies correctly set, so we can remove.

File: var/spack/repos/builtin/packages/sim-model/package.py (hunk 22 above)
Comment (ferdonline) [reply var/spack/repos/builtin/packages/sim-model/package.py]: we could use a plain loop, it's a matter of preference (notice the `for` inside the join)

File: var/spack/repos/builtin/packages/sim-model/package.py (hunk 23 above)
Comment (ferdonline) [reply var/spack/repos/builtin/packages/sim-model/package.py]: to avoid creating unwanted env vars

File: var/spack/repos/builtin/packages/sim-model/package.py (hunk 24 above)
Comment (ferdonline) [reply var/spack/repos/builtin/packages/sim-model/package.py:180]: yep, by design we should support it. neurodamus-py already supports multiple libs

File: var/spack/repos/builtin/packages/model-neocortex/package.py (hunk 9 above)
Comment (ferdonline) [reply var/spack/repos/builtin/packages/model-neocortex/package.py:10]: We should, however I don't like the idea of replicating all the code though (which is almost 100% copy-paste from neurodamus-neocortex), so I thought for now we offer this one for people to play and later do it in a better way

File: var/spack/repos/builtin/packages/highfive/package.py (hunk 5 above)
Comment (ferdonline) [reply var/spack/repos/builtin/packages/highfive/package.py:45]: Just checking, it seems both ways are used interchangeably... 
```
leite@bluebrain291 packages $ grep 'in self.spec' * -R | wc -l
     537
leite@bluebrain291 packages $  grep 'spec.satisfies' * -R | wc -l
     472
leite@bluebrain291 packages $ grep "'TRUE' if" * -R | wc -l
       9
```
Now between
`'-DUSE_EIGEN:Bool=' + str(self.spec.satisfies('+eigen'))`
and
`'-DUSE_EIGEN:Bool=' + ('TRUE' if '+eigen' in self.spec else 'FALSE')`
Do you have a strong preference? IMHO +1 for the shortest and the `satisfies` API :)

File: var/spack/repos/builtin/packages/sim-model/package.py (hunk 12 above)
Comment (ferdonline) [reply var/spack/repos/builtin/packages/sim-model/package.py]: new commit changes to to_rm

File: deploy/configs/applications/modules.yaml (hunk 3 above)
Comment (pramodk) [reply deploy/configs/applications/modules.yaml:43]: @ferdonline: ok...what I meant was can we add +common or -common there in the module name so that it’s clear.

File: var/spack/repos/builtin/packages/highfive/package.py (hunk 5 above)
Comment (pramodk) [reply var/spack/repos/builtin/packages/highfive/package.py:45]: It’s ok.
For the second, I have always seen later.

File: var/spack/repos/builtin/packages/neurodamus-core/package.py (hunk 14 above)
Comment (pramodk) [reply var/spack/repos/builtin/packages/neurodamus-core/package.py:52]: In that case ‘libs’ property of hdf5 should provide necessary libraries? 
Hmm...but I see the possible issues there...it’s ok as it is.

File: var/spack/repos/builtin/packages/neurodamus-model/package.py (hunk 16 above)
Comment (pramodk) [reply var/spack/repos/builtin/packages/neurodamus-model/package.py]: But for final exe .i.e. special we link with MPI wrappers through NEURON. So I will just say depends_on(‘mpi’).

Comment (ferdonline) [reply var/spack/repos/builtin/packages/neurodamus-model/package.py]: I sincerely had it like that, but `spec['mpi'].libs` failed (at build time, saying it cannot find the libraries). And since we use mpicc, which do the linking for us, I don't see a better alternative.

File: var/spack/repos/builtin/packages/sim-model/package.py (hunk 22 above)
Comment (ferdonline) [reply var/spack/repos/builtin/packages/sim-model/package.py]: Changed to a plain loop

File: var/spack/repos/builtin/packages/sim-model/package.py (hunk 18 above)
Comment (ferdonline) [reply var/spack/repos/builtin/packages/sim-model/package.py:36]: I added a comment to explain well why we dont set link mode.

File: var/spack/repos/builtin/packages/neurodamus-model/package.py (hunk 17 above)
Comment (ferdonline) [reply var/spack/repos/builtin/packages/neurodamus-model/package.py:136]: we are not sure about the order the libs are found

**Comment by ferdonline** [comment]
Simulation Stack run:
https://bbpcode.epfl.ch/ci/blue/organizations/jenkins/hpc.SimulationStack/detail/hpc.SimulationStack/1052/pipeline/154

File: deploy/configs/applications/modules.yaml [hunk 25]
```diff
@@ -83,6 +83,7 @@ modules:
         '^python@3:': '/python3'
         '+mpi': '/parallel'
         '~mpi': '/serial'
+        '+common': '+common'  # neurodamus-core
```
Comment (pramodk) [inline deploy/configs/applications/modules.yaml]: Are we deploying two modules of neurodamus-core? are both as modules required?

By the way to avoid error as serial/parallel, it should be : 
```suggestion
        '^neurodamus-core+common': '+common'  # neurodamus-core
```

(I don't expect packages with variant +common but still...)

**Review by pramodk — DISMISSED** [review]

File: deploy/configs/applications/modules.yaml (hunk 25 above)
Comment (ferdonline) [reply deploy/configs/applications/modules.yaml]: > Are we deploying two modules of neurodamus-core? are both as modules required?

Yep if we still want to support the case Werner had.

would `^neurodamus-core+common` work? would be nice but neurodamus-core is the end package

**Comment by ferdonline** [comment]
@pramodk I have checked the modules that are created. It seems ok now, but nevertheless I still dont understand well why `neurodamus-core+common` gets three modules
```
neurodamus-core/2.7.3/python3/parallel/serial-commonmods
neurodamus-core/2.7.3/python3/parallel-commonmods
neurodamus-core/2.9.1/python3/parallel/serial
neurodamus-core/2.9.1/python3/parallel/serial-commonmods
```
My intention was to have only
```
neurodamus-core/2.7.3/python3/parallel/serial-commonmods 
neurodamus-core/2.9.1/python3/parallel/serial-commonmods 
neurodamus-core/2.9.1/python3/parallel/serial 
```
Anyway it's a detail which I am not sure in production will look like this.
If you want to make some changes, please go ahead. I'm happy with it and I think we have to close this soon

**Comment by ferdonline** [comment]
Oh gosh, now it created 10 modules for synapsetool... 😞 smtg is not ok

File: deploy/packages/bbp-packages.yaml [hunk 26]
```diff
@@ -74,7 +74,7 @@ packages:
     specs:
       - functionalizer@3.12.2
       - parquet-converters@0.5.2
-      - synapsetool@0.5.6
+      - synapsetool
```
Comment (matz-e) [inline deploy/packages/bbp-packages.yaml]: I think this creates the `synapsetool` modules… better to pin the latest version.

Comment (ferdonline) [reply deploy/packages/bbp-packages.yaml]: I thought this was asking to install last version. How does it work then?

**Review by matz-e — APPROVED** [review]

**Comment by pramodk** [comment]
I am just sick of seeing ‘parallel/serial’ in module name...
@matz-e : please do something!

**Comment by ferdonline** [comment]
Latest SimStack green: https://bbpcode.epfl.ch/ci/blue/organizations/jenkins/hpc.SimulationStack/detail/hpc.SimulationStack/1055/pipeline/46

**Comment by matz-e** [comment]
> I am just sick of seeing ‘parallel/serial’ in module name...
> @matz-e : please do something!

Will require a maintenance window, to also get rid of the `/` and replace them with `-`.

**Comment by ferdonline** [comment]
> rid of the / and replace them with -

Notice that `-` is the default separator. It's enough to do `'+mpi': 'parallel'`

**Comment by matz-e** [comment]
> > rid of the / and replace them with -
> 
> Notice that `-` is the default separator. It's enough to do `'+mpi': 'parallel'`

We have some special case in the code that took the `/` into account… forgot how it works, but it's gone in the new deployment, where `-` is standard.
